esx_service/vmdk_ops.py

```diff
@@ -1075,22 +1084,115 @@ def load_vmci():
    else:
        lib = CDLL(os.path.join(LIB_LOC, "libvmci_srv.so"), use_errno=True)
 
+def getLock(lockname):
+    '''
+    Create or return a existing lock identified by lockname.
+    '''
+    global rsrcLocks
+    global managerLock
+
+    if not rsrcLocks:
+        rsrcLocks = WeakValueDictionary()
+        logging.debug("getLock(): Initialized Lock store")
+    if not managerLock:
+        managerLock = threading.Lock()
+        logging.debug("getLock(): Initialized Manager Lock")
+    
+    with managerLock:
+        try:
+            lock = rsrcLocks[lockname]
+            logging.debug("getLock(): return existing lock %s", lockname)
+        except KeyError as e:
+            lock = threading.Lock()
+            rsrcLocks[lockname] = lock
+            logging.debug("getLock(): return new lock %s", lockname)
+    return lock
+
+def execRequestThread(clntsock, cartel, request):
+    '''
+    Execute requests in a thread context with a per volume locking. 
+    '''
+    # Before we start, block to allow main thread to advance.
+    time.sleep(0.001)
```

Comment:
Yes, there's no guarantee.

It is a "best effort" to allow the main thread (or other threads) to advance to the next request  because we will start executing bytecode and only after we block, everything else will have opportunity to schedule. The python global interpreter lock does not help here, and this kind of trickery is even recommended in the python documentation.

A semaphore here wouldn't have the same effect plus the overhead? We'll probably have more threads blocking on acquire(), and the order or priority of waking cannot be relied on.